src/server/bitops_family.cc

```diff
@@ -181,60 +253,237 @@ bool SetBitValue(uint32_t offset, bool bit_value, std::string* entry) {
 
-  if (added) {  // this is a new entry in the "table"
+  auto find_res = element_access.Find(shard);
+
+  if (find_res != OpStatus::OK) {
+    return find_res;
+  }
+
+  if (element_access.IsNewEntry()) {
     std::string new_entry(GetByteIndex(offset) + 1, 0);
     old_value = SetBitValue(offset, bit_value, &new_entry);
-    UpdateBitMapValue(new_entry);
+    element_access.Commit(new_entry);
   } else {
-    if (it->second.ObjType() != OBJ_STRING) {
-      return OpStatus::WRONG_TYPE;
-    }
     bool reset = false;
-    std::string existing_entry{GetString(args_.shard, it->second)};
-    if ((existing_entry.size() * OFFSET_FACTOR) <= offset) {  // need to resize first
+    std::string existing_entry{element_access.Value()};
+    if ((existing_entry.size() * OFFSET_FACTOR) <= offset) {
       existing_entry.resize(GetByteIndex(offset) + 1, 0);
       reset = true;
     }
     old_value = SetBitValue(offset, bit_value, &existing_entry);
     if (reset || old_value != bit_value) {  // we made a "real" change to the entry, save it
-      UpdateBitMapValue(existing_entry);
+      element_access.Commit(existing_entry);
     }
   }
   return old_value;
 }
 
+// ---------------------------------------------------------
+
+std::string RunBitOperationOnValues(std::string_view op, const BitsStrVec& values) {
+  // This function accept an operation (either OR, XOR, NOT or OR), and run bit operation
+  // on all the values we got from the database. Note that in case that one of the values
+  // is shorter than the other it would return a 0 and the operation would continue
+  // until we ran the longest value. The function will return the resulting new value
+  std::size_t max_len = 0;
+  std::size_t max_len_index = 0;
+
+  const auto BitOperation = [&]() {
+    if (op == OR_OP_NAME) {
+      std::string default_str{values[max_len_index]};
+      return BitOpString(OrOp, SkipOr, std::move(values), std::move(default_str));
+    } else if (op == XOR_OP_NAME) {
+      return BitOpString(XorOp, SkipXor, std::move(values), std::string(max_len, 0));
+    } else if (op == AND_OP_NAME) {
+      return BitOpString(AndOp, SkipAnd, std::move(values), std::string(max_len, 0));
+    } else if (op == NOT_OP_NAME) {
+      return BitOpNotString(values[0]);
+    } else {
+      LOG(FATAL) << "Operation not supported '" << op << "'";
+      return std::string{};  // otherwise we will have warning of not returning value
+    }
+  };
+
+  if (values.empty()) {  // this is ok in case we don't have the src keys
+    return std::string{};
+  }
+  // The new result is the max length input
+  max_len = values[0].size();
+  for (std::size_t i = 1; i < values.size(); ++i) {
+    if (values[i].size() > max_len) {
+      max_len = values[i].size();
+      max_len_index = i;
+    }
+  }
+  return BitOperation();
+}
+
+OpResult<std::string> CombineResultOp(ShardStringResults result, std::string_view op) {
+  // take valid result for each shard
+  BitsStrVec values;
+  int index = 0;
```

Comment:
why do you need `index` ?

Reply:
for the first case, changed the code that call this function, now for any key that was not found, we will not set anything, so this can be empty.
For the second issue, remove (I think it was leftover from the refactor).